src/mongo/delegates/Views.class.php

```diff
@@ -532,6 +476,109 @@ public function generateView($viewId,$resource=null,$context=null,$queueName=nul
         }
     }
 
+    protected function ensureIndices($viewSpec, $collection) {
+
+        // ensure that the ID field, view type, and the impactIndex indexes are correctly set up
+        $collection->ensureIndex(
+            array(
+                '_id.r'=>1,
+                '_id.c'=>1,
+                '_id.type'=>1
+            ),
+            array(
+                'background'=>1
+            )
+        );
+
+        $collection->ensureIndex(
+            array(
+                '_id.type'=>1
+            ),
+            array(
+                'background'=>1
+            )
+        );
+
+        $collection->ensureIndex(
+            array(
+                'value.'._IMPACT_INDEX=>1
+            ),
+            array(
+                'background'=>1
+            )
+        );
+
+        // spec type/revision index lets us find outdated views
+        Config::ensureIndexForSpecRevision($collection);
+
+        // ensure any custom view indexes
+        if (isset($viewSpec['ensureIndexes']))
+        {
+            foreach ($viewSpec['ensureIndexes'] as $ensureIndex)
+            {
+                $collection->ensureIndex(
+                    $ensureIndex,
+                    array(
+                        'background'=>1
```

Comment:
Should these be sparse indexes?  If they index properties that are specific to a subset of documents, it seems wasteful to index everything.